### ISSUE TITLE
Losing metadata when updating an object implementing Persistable

### ISSUE THREAD

**Issue by gquemener** [mongodb/mongo-php-library#104]
Hello,

I'm having trouble updating one document that implements the Persistable interface.
It's unserialization metadata get lost every time I replace the document.

Here's an exemple showing the problem:

``` php
<?php

require 'vendor/autoload.php';

class A implements \MongoDB\BSON\Persistable
{
    private $name;
    public function __construct($name)
    {
        $this->name = $name;
    }

    public function setName($name)
    {
        $this->name = $name;
    }

    public function bsonSerialize()
    {
        return [
            'name' => $this->name
        ];
    }

    public function bsonUnserialize(array $data)
    {
        $this->name = $data['name'];
    }
}

$instance = new A('foo');

$manager = new \MongoDB\Driver\Manager('mongodb://mongo:27017');
$collection = new \MongoDB\Collection($manager, 'test', 'coll');
$result = $collection->insertOne($instance);

/**
 * > db.coll.find()
 * {
 *         "_id" : ObjectId("56a269e0046d5f0014645471"),
 *         "__pclass" : BinData(128,"QQ=="),
 *         "name" : "foo"
 * }
 */

$instance->setName('bar');
$collection->replaceOne(['_id' => $result->getInsertedId()], $instance);

/**
 * > db.coll.find()
 * { "_id" : ObjectId("56a26a97046d5f0039107a21"), "name" : "bar" }
 */
```

What am I doing wrong here please?

I'm running mongodb server v3.2.1.

Thanks in advance!

**Comment by gquemener** [comment]
Hello!

I've found a way to do it, but I'm not sure this is the correct one:
``` php
$instance->setName('bar');
$collection->updateOne(['_id' => $result->getInsertedId()], [
    '$set' => $instance->bsonSerialize()
]);
```

What do you think?

**Comment by gquemener** [comment]
No after research, this is definitly not a good idea. `bsonSerialize()` is used internally by the MongoDB driver. Anyway it doesn't work with embedded document.

**Comment by jmikola** [comment]
I believe the problem is due to your class not handling the `_id` field.

When inserting the instance, `bsonSerialize()` returns a document containing only a `name` field, which leaves the driver to generate an ObjectId on its own (the same that you later retrieve from the write result). When replacing the document, `bsonSerialize()` is again invoked and returns another document containing only `name` but with a different value. In my local test, that replacement does achieve the desired goal of changing the `name` field of the inserted document. I verified this using the shell to dump the document after the `insertOne()` and `replaceOne()` calls and there is but one document in the collection:

```
> db.coll.find()
{ "_id" : ObjectId("56a6a61c6118fd39a911c6e1"), "name" : "bar" }
```

I would suggest the following to properly handle the `_id` property on your class:

 * Add an `$_id` property on the class (or whatever you'd like the name it)
 * Ensure the property is included in `bsonSerialize()` if set. Unless you plan to initialize `$_id` to a new ObjectId (or some other value) during construction, I'd suggest you *not* include it in `bsonSerialize()` if it's `null`, as that will happy create a document in MongoDB with a `null` ID value (and naturally start throwing duplicate key exceptions after the first insertion).
 * Ensure the property is picked up in `bsonUnserialize()`.

For example:

```
class A implements MongoDB\BSON\Persistable
{
    private $_id;
    private $name;

    public function __construct($name)
    {
        $this->name = $name;
    }

    public function setName($name)
    {
        $this->name = $name;
    }

    public function bsonSerialize()
    {
        $data = [
            'name' => $this->name,
        ];

        if ($this->_id !== null) {
            $data['_id'] = $this->_id;
        }

        return $data;
    }

    public function bsonUnserialize(array $data)
    {
        $this->name = $data['name'];
        $this->_id = $data['_id'];
    }
}
```

That said, I think it would be easiest to assign a new `MongoDB\BSON\ObjectID` to `$this->_id` in your constructor and simplify `bsonSerialize()`. I expect this may end up being a common mistake for new users, so I will get this added to our documentation examples soon (both on PHP.net and in the upcoming library documentation for using Persistable and the other two interfaces).

----

That said, there is a driver problem here (below the library). The `__pclass` property is disappearing after the replacement is executed. Curiously, it is present when we debug the replacement document through the BSON serialization functions:

```
echo MongoDB\BSON\toJSON(MongoDB\BSON\fromPHP($instance));
// outputs: { "__pclass" : { "$type" : "80", "$binary" : "QQ==" }, "name" : "bar" }
```

Looking into the extension code, I believe we're missing `PHONGO_BSON_ADD_ODS|PHONGO_BSON_ADD_CHILD_ODS` flags when encoding the replacement arguments in [BulkWrite.c](https://github.com/mongodb/mongo-php-driver/blob/94da5474070ad80b7e69da41c720d17a73cb94b4/src/MongoDB/BulkWrite.c#L153). Those are flags we use when inserting documents, and I believe missing them for replacements was an oversight. I'll look into that and cross-reference this with a JIRA ticket on the extension soon.

Thanks for the report!

**Comment by jmikola** [comment]
Cross-referencing:

 * https://jira.mongodb.org/browse/PHPC-549 for documentation improvements
 * https://jira.mongodb.org/browse/PHPC-550 for replacement doc encoding bug

**Comment by gquemener** [comment]
FYI, while waiting for the fix, one must specificaly serialize the `__pclass` property to prevent metadata to be le lost on document update:

```php
    public function bsonSerialize()
    {
        return [
            '_id' => $this->id,
            '__pclass' => new BSON\Binary(get_class($this), BSON\Binary::TYPE_USER_DEFINED),
            //... whatever other fields you want to persist
        ];
    }
```

**Comment by jmikola** [comment]
FYI: [PHPC-550](https://jira.mongodb.org/browse/PHPC-550) has been merged and is slated for a 1.1.3 release this week.

**Comment by jmikola** [comment]
[1.1.3](https://github.com/mongodb/mongo-php-driver/releases/tag/1.1.3) has been released, so I think this can be closed. Feel free to follow up if the issue persists.
